### PR TITLE
SOLR-8306: Enhance ExpandComponent to allow expand.hits=0

### PR DESCRIPTION
**Pull request by ameliahenderson**
# Description

Enables ExpandComponent to allow expand.hits=0 for those who don't want an expanded document returned and only want the numFound returned from the expand section.

# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

We have been using this patch in our internal fork of SOLR for years so this has been thoroughly tested to ensure this functionality works as desired. Unit tests have been added for this functionality and all tests pass.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `master` branch.
- [x] I have run `ant precommit` and the appropriate test suite.
- [x] I have added tests for my changes.
- [x] I have added documentation for the [Ref Guide](https://github.com/apache/lucene-solr/tree/master/solr/solr-ref-guide) (for Solr changes only).

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/handler/component/ExpandComponent.java [hunk 1]
```diff
@@ -668,6 +676,18 @@ public void collect(int docId) throws IOException {
 
   }
 
+  private static Collector getExpandCollector(int limit, Sort sort) throws IOException {
```
Comment (munendrasn) [inline solr/core/src/java/org/apache/solr/handler/component/ExpandComponent.java]: I think this method should be part of `GroupCollector` (default implementation) and method name could simply be `getCollector`

File: solr/core/src/java/org/apache/solr/search/DocSlice.java [hunk 2]
```diff
@@ -56,7 +56,7 @@ public DocSlice(int offset, int len, int[] docs, float[] scores, long matches, f
     this.scores=scores;
     this.matches=matches;
     this.maxScore=maxScore;
-    this.ramBytesUsed = BASE_RAM_BYTES_USED + ((long)docs.length << 2) + (scores == null ? 0 : ((long)scores.length<<2)+RamUsageEstimator.NUM_BYTES_ARRAY_HEADER);
+    this.ramBytesUsed = BASE_RAM_BYTES_USED + (docs == null ? 0 : ((long)docs.length << 2)) + (scores == null ? 0 : ((long)scores.length<<2)+RamUsageEstimator.NUM_BYTES_ARRAY_HEADER);
```
Comment (munendrasn) [inline solr/core/src/java/org/apache/solr/search/DocSlice.java]: 👍

File: solr/core/src/test/org/apache/solr/handler/component/DistributedExpandComponentTest.java [hunk 3]
```diff
@@ -153,6 +153,23 @@ public void test() throws Exception {
     assertExpandGroupCountAndOrder("group3", 1, results, "9");
     assertExpandGroupCountAndOrder("group4", 1, results, "14");
 
+    //Test expand.rows = 0 - no docs only expand count
+
+    params = new ModifiableSolrParams();
```
Comment (munendrasn) [inline solr/core/src/test/org/apache/solr/handler/component/DistributedExpandComponentTest.java]: Is it possible to add a case w/o collapse too (with expand.field) and with fl with and w/o `score`?

File: solr/core/src/test/org/apache/solr/handler/component/TestExpandComponent.java [hunk 4]
```diff
@@ -234,6 +234,21 @@ private void _testExpand(String group, String floatAppend, String hint) throws E
         "/response/lst[@name='expanded']/result[@name='2"+floatAppend+"']/doc[1]/str[@name='id'][.='8']"
     );
 
+    //Test expand.rows = 0 - no docs only expand count
+    params = new ModifiableSolrParams();
```
Comment (munendrasn) [inline solr/core/src/test/org/apache/solr/handler/component/TestExpandComponent.java]: Same as above so that we can cover more cases

Comment (ameliahenderson) [reply solr/core/src/test/org/apache/solr/handler/component/TestExpandComponent.java]: done

File: solr/core/src/test/org/apache/solr/handler/component/DistributedExpandComponentTest.java (hunk 3 above)
Comment (ameliahenderson) [reply solr/core/src/test/org/apache/solr/handler/component/DistributedExpandComponentTest.java]: done

File: solr/core/src/java/org/apache/solr/handler/component/ExpandComponent.java (hunk 1 above)
Comment (ameliahenderson) [reply solr/core/src/java/org/apache/solr/handler/component/ExpandComponent.java]: done

File: solr/CHANGES.txt [hunk 5]
```diff
@@ -57,7 +57,7 @@ Improvements
 
 Optimizations
 ---------------------
-(No changes)
+* SOLR-8306: Enhance ExpandComponent to allow expand.hits=0 (Marshall Sanders, Amelia Henderson)
```
Comment (munendrasn) [inline solr/CHANGES.txt]: Maybe change the wording here, something like `Do not collect expand documents when expand.rows=0`
there is no `expand.hits` parameter

File: solr/solr-ref-guide/src/collapse-and-expand-results.adoc [hunk 6]
```diff
@@ -142,6 +142,8 @@ The “expand=true” parameter turns on the ExpandComponent. The ExpandComponen
 
 Inside the expanded section there is a _map_ with each group head pointing to the expanded documents that are within the group. As applications iterate the main collapsed result set, they can access the _expanded_ map to retrieve the expanded groups.
 
+If you would like to have only the number of documents found from the groups in the expanded section returned and not the actual documents, you can set "expand.rows=0". Note that when using "expand.rows=0", score won’t be computed for the expanded section even when requested so maxScore will not be available.
```
Comment (munendrasn) [inline solr/solr-ref-guide/src/collapse-and-expand-results.adoc]: Thanks for the documentation update. I think this should be part of `expand.rows=0`. Something like this https://github.com/apache/lucene-solr/pull/1334/files#diff-9c5999131702feec07a6b7057fa3bcd7R23

Comment (munendrasn) [reply solr/solr-ref-guide/src/collapse-and-expand-results.adoc]: Example 
```
[IMPORTANT]
====
When `expand.rows=0`, then only the number of documents found for each expanded value is returned. Hence, scores won't be computed even if requested. `maxScore` is set to 0
====
```

File: solr/CHANGES.txt (hunk 5 above)
Comment (ameliahenderson) [reply solr/CHANGES.txt]: Thanks for that. updated

File: solr/solr-ref-guide/src/collapse-and-expand-results.adoc (hunk 6 above)
Comment (ameliahenderson) [reply solr/solr-ref-guide/src/collapse-and-expand-results.adoc]: Updated. Please let me know if it's still in wrong place

**Comment by munendrasn** [comment]
I had to rebase with master due to some conflicts

**Comment by ameliahenderson** [comment]
thanks for taking care of rebase and uploading patch again

**Pull request merged**